### PR TITLE
Improvements to struct field completion

### PR DESCRIPTION
**Pull request by lukaszsamson**
With this PR elixir-ls is able to provide completions for map keys and struct fields in following cases
```
struct = %My{}
struct.f
```

```
struct = %My{}
%{struct | f
```

```
map = %{key: 123}
%{map | k
```

```
map = %{key: 123, nested: %{module: String}}
map.k
map.nested.m
map.nested.module.pri
```

```
atom = String
string.pri
```

### CODE REVIEW

**Comment by axelson** [comment]
Wow this looks like a really neat change! It would be nice if we could add some tests to go along with this (probably on the provider level instead of the server level).

**Comment by axelson** [comment]
Might be nice to add some docs to the Readme as well. That way users can know when to expect completion of struct fields to work.

**Comment by lukaszsamson** [comment]
@axelson I added tests and a doc section in readme. I also fixed a small LSP spec compliance bug  in https://github.com/elixir-lsp/elixir-ls/pull/202/commits/fa69c6c37d654171f293bfbc472e9228ea3d0083

File: README.md [hunk 1]
```diff
@@ -87,6 +87,27 @@ You can control which warnings are shown using the `elixirLS.dialyzerWarnOpts` s
 
 ElixirLS's Dialyzer integration uses internal, undocumented Dialyzer APIs, and so it won't be robust against changes to these APIs in future Erlang versions.
 
+## Code completion
+
+ElixirLS bundles an advanced code completion provider. The provider builds on [Elixir Sense](https://github.com/elixir-lsp/elixir_sense) library and utilizes two main mechanisms. The first one is reflection - getting information about compiled modules by Erlang and Elixir APIs. The second one is AST analysis of the current text buffer. While reflection gives precise results, it is not well suited for on demand completion of symbols from the currently edited file. The compiled version is likely to be outdated or the file may not compile at all. AST analysis helps in that case but it has its limitations. Unfortunately it is infeasible to be 100% accurate, especially with Elixir being a metaprogramming heavy language.
```
Comment (axelson) [inline README.md:92]:
```suggestion
ElixirLS bundles an advanced code completion provider. The provider builds on [Elixir Sense](https://github.com/elixir-lsp/elixir_sense) library and utilizes two main mechanisms. The first one is reflection - getting information about compiled modules from Erlang and Elixir APIs. The second one is AST analysis of the current text buffer. While reflection gives precise results, it is not well suited for on demand completion of symbols from the currently edited file. The compiled version is likely to be outdated or the file may not compile at all. AST analysis helps in that case but it has its limitations. Unfortunately it is infeasible to be 100% accurate, especially with Elixir being a metaprogramming heavy language.
```

File: README.md [hunk 2]
```diff
@@ -87,6 +87,27 @@ You can control which warnings are shown using the `elixirLS.dialyzerWarnOpts` s
...
+
+The completions include:
+
+- keywords
+- special form snippets
+- functions
+- macros
+- modules
+- variables
+- struct fields (if the struct type is explicitely stated or can be infered from variable binding)
```
Comment (axelson) [inline README.md:102]:
```suggestion
- struct fields (only if the struct type is explicitly stated or can be inferred from the variable binding)
```

File: README.md [hunk 3]
```diff
@@ -87,6 +87,27 @@ You can control which warnings are shown using the `elixirLS.dialyzerWarnOpts` s
...
+- atom map keys (if map keys can be infered from variable binding)
+- attrubutes
```
Comment (axelson) [inline README.md:104]:
```suggestion
- attributes
```

**Comment by lukaszsamson** [comment]
Good catch @axelson, fixed in elixir_sense and dep updated.

> Additionally I don't think we should return the __struct__ field as a completion. Maybe we should filter it out on the ElixirLS side?

I've got mixed feelings about that. On one hand underscored symbols are meant to be hidden but on the other hand I happen to use them quite often. Similar issue applies to functions, macros, modules and types with `@doc false`. Should we hide them as well? In my opinion not. It's generally advisable to stick to public APIs while using libraries but when writing libraries you work mostly with your own private APIs.

Maybe we should just deprioritise private API completions and annotate them as such. If you agree, let's do it in another PR.

**Comment by axelson** [comment]
> I've got mixed feelings about that. On one hand underscored symbols are meant to be hidden but on the other hand I happen to use them quite often. Similar issue applies to functions, macros, modules and types with @doc false. Should we hide them as well? In my opinion not. It's generally advisable to stick to public APIs while using libraries but when writing libraries you work mostly with your own private APIs.

I do have general feelings to you on private/hidden symbols. But for this in particular I'm suggesting to specifically hide `__struct__` since I think users will very often be completing struct fields, and almost never want to complete the `__struct__` field. i.e. code like `a = %MyStruct{}`; struct = a.__struct__`. Instead, they could use pattern matching like `%struct{} = %MyStruct{}` (but that's also very rarely needed).

> Maybe we should just deprioritise private API completions and annotate them as such. If you agree, let's do it in another PR.

Agreed and agreed. Filed as #214 

I'm going to run with this locally for a bit longer and see if I encounter any further issues.

**Pull request merged**